org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/framework/protobuf/ProtobufSupport.java

```diff
@@ -0,0 +1,195 @@
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.UUID;
+import java.util.stream.Stream;
+
+import org.apache.commons.lang3.StringUtils;
+import org.eclipse.buildship.core.BuildConfiguration;
+import org.eclipse.buildship.core.GradleBuild;
+import org.eclipse.buildship.core.GradleCore;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.jdt.core.IClasspathAttribute;
+import org.eclipse.jdt.core.IClasspathEntry;
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.core.JavaCore;
+import org.eclipse.jdt.core.JavaModelException;
+import org.eclipse.jdt.ls.core.internal.ActionableNotification;
+import org.eclipse.jdt.ls.core.internal.JavaClientConnection.JavaLanguageClient;
+import org.eclipse.jdt.ls.core.internal.JavaLanguageServerPlugin;
+import org.eclipse.jdt.ls.core.internal.ProgressReport;
+import org.eclipse.jdt.ls.core.internal.ProjectUtils;
+import org.eclipse.jdt.ls.core.internal.framework.IFrameworkSupport;
+import org.eclipse.jdt.ls.core.internal.managers.GradleProjectImporter;
+import org.eclipse.jdt.ls.core.internal.preferences.PreferenceManager;
+import org.eclipse.lsp4j.Command;
+import org.eclipse.lsp4j.MessageType;
+
+public class ProtobufSupport implements IFrameworkSupport {
+
+	/**
+	 * The classpath entry attribute which will be added for the protobuf source output directories.
+	 */
+	public static final String PROTOBUF_GENERATED_SOURCE = "protobuf_generated_source";
+
+	@Override
+	public void onDidProjectsImported(IProgressMonitor monitor) {
+		PreferenceManager preferencesManager = JavaLanguageServerPlugin.getPreferencesManager();
+		if (preferencesManager == null) {
+			return;
+		}
+
+		if (!preferencesManager.getPreferences().isProtobufSupportEnabled()) {
+			return;
+		}
+
+		List<IProject> projects = ProjectUtils.getGradleProjects();
+		if (projects.isEmpty()) {
+			return;
+		}
+
+		List<Object> projectUris = new ArrayList<>();
+		for (IProject project : projects) {
+			if (!ProjectUtils.isJavaProject(project)) {
+				continue;
+			}
+
+			Set<File> protobufOutputDirs = findProtobufOutputDirectories(project);
+			if (protobufOutputDirs.isEmpty()) {
+				continue;
+			}
+
+			boolean hasGenerated = containsJavaFiles(protobufOutputDirs);
+			if (!hasGenerated) {
+				projectUris.add(project.getLocationURI().toString());
+			}
+		}
+		if (!projectUris.isEmpty()) {
+			ActionableNotification notification = new ActionableNotification().withSeverity(MessageType.Info)
+					.withMessage("Would you like to generate Java source files out of your proto files?")
+					.withCommands(Arrays.asList(new Command("Yes", "java.protobuf.generateSources", Arrays.asList(projectUris))));
```

Comment:
@jdneo I've tested the change and it works well. Apologies for taking long. Regarding the prompt, I would change "Yes" to "Generate". I think if we have a Yes/No dialog but we only want to display the action that does something, then we can just name it by the action that will be taken.

Reply:
`Generate` sounds better than `Yes`. 👍

I'll raise another PR to change it